### ISSUE TITLE
Proposal for output format

### ISSUE THREAD

**Issue by frgomes** [frgomes/scala-bindgen#12]
## Proposal (this is subject to change)

Given input file ``getopt.h``, we should generate something looking like down below.

Just some initial considerations:
* all parameters are passed by value
* the syntax may or may not be implemented, in particular annotations.

```scala
object getopt {
  
  @struct
  case class option(... to be investigated ...)

  object api extends api

  @link("getopt")
  trait api {
    var optarg: Ptr[CString] = extern
    var optind: Int = extern
    var opterr: Int = extern
    var optopt: Int = extern

    def getopt
                argc: Int, 
                argv: Array[@const Ptr[CString]], 
                optstring: Array[@const Ptr[CString]]): Int = extern

    def getopt_long(
                argc: Int, 
                argv: Array[@const Ptr[CString]], 
                optstring: Array[@const Ptr[CString]], 
                longopts: Ptr[@const option], 
                longindex: Ptr[Int]): Int = extern

    def getopt_long_only(
                argc: Int, 
                argv: Array[@const Ptr[CString]], 
                optstring: Array[@const Ptr[CString]], 
                longopts: Ptr[@const option], 
                longindex: Ptr[Int]): Int = extern
}
```

**Comment by frgomes** [comment]
I'm going to update this proposal (without notice!) in order to comply with
https://github.com/scala-native/scala-native/issues/897

**Comment by frgomes** [comment]
@jonas I don't like this proposal above sufficiently enough. It leaves opportunities for name clash. If you have on the C side something called ``api``, it will clash with ``object api extends api``.

Maybe a C module ``getopt`` could end up on something like this:

```scala
object getopt extends getopt
trait getopt {

  trait defines {
    @struct
    case class option(... to be investigated ...)
  }

  object externs extends externs

  @link("geopt")
  trait externs {
      var optarg: Ptr[CString] = extern
      var optind: Int = extern
      var opterr: Int = extern
      var optopt: Int = extern

      def getopt
                argc: Int, 
                argv: Array[@const Ptr[CString]], 
                optstring: Array[@const Ptr[CString]]): Int = extern

      def getopt_long(
                argc: Int, 
                argv: Array[@const Ptr[CString]], 
                optstring: Array[@const Ptr[CString]], 
                longopts: Ptr[@const defines#option],  
                longindex: Ptr[Int]): Int = extern
  }
}
```

**Comment by jonas** [comment]
What's the reason for nesting the `extern` declarations in a trait or object?

**Comment by frgomes** [comment]
@jonas : This comes from the previous way of doing things. If I'm not mistaken, ``@link`` was associated to an ``object`` which aggregated all ``extern`` definitions. So, I invented ``trait defines`` and ``trait externs`` in order to separate glorified constants, typedefs, etc... everything zero-sized from symbols which can be effectively linked.

I've just seen that spec 2.0 says that ``@link`` can appear to any trait, method, field, etc. So, we don't really need these additional traits. But I still think that "it looks nice and organized". I think that adding ``@link`` to every field or method would look confusing and verbose.

Ideas?

**Comment by frgomes** [comment]
@jonas : Open for voting.

```scala
object getopt extends getopt
trait getopt {

  @struct
  case class option(... to be investigated ...)

  @link("geopt")
  var optarg: Ptr[CString] = extern


  @link("geopt")
  var optind: Int = extern

  @link("geopt")
  var opterr: Int = extern

  @link("geopt")
  var optopt: Int = extern

  @link("geopt")
  def getopt
            argc: Int, 
            argv: Array[@const Ptr[CString]], 
            optstring: Array[@const Ptr[CString]]): Int = extern

  @link("geopt")
  def getopt_long(
            argc: Int, 
            argv: Array[@const Ptr[CString]], 
            optstring: Array[@const Ptr[CString]], 
            longopts: Ptr[@const defines#option],  
            longindex: Ptr[Int]): Int = extern

}
```

**Comment by jonas** [comment]
I'm personally a fan of the latest proposal. It is flatter and matches the hand-written styles in Scala Native.

In case we start generating implicits for certain APIs, we definitely should put them in a separate module to make them opt-in.

We should probably also consider using packages if generating for multiple header files. For example for [libuv](https://github.com/libuv/libuv/tree/v1.x/include) we might want to generate package such as:

```scala
import uv.{run, stop}
import uv.posix._
import uv.threadpool._
```

Getting a bit carried away here though.

**Comment by frgomes** [comment]
OK. Let's take the last approach then.
Multiple files is just around the corner. It's a very common use case. See #13

**Comment by frgomes** [comment]
@jonas : Can you explain your idea regarding packages?

**Comment by jonas** [comment]
The idea is to generate separate scala modules or packages for each header file keeping the original structure. One benefit is that the original documentation more naturally maps to the Scala bindings.

**Comment by frgomes** [comment]
@jonas : There's already a command line option for specifying the root package, but we are not using it at the moment. Also, there are considerations regarding ``#include`` directives which we need to take into account. Suppose the command below:
```bash
$ scala-bindgen app.h -C /path/to/sources --package com.example.app -- -I /path/to/my/includes
```
and imagine ``app.h`` has:
```c
#include <sys/unistd.h> // located in system includes
#include "shapes.h" // located in /path/to/my/includes
#include "codes.h" // located alongside to this file
```

There are several details to consider: (these items are briefly discussed at #23)
1. How deep recursively we should dive into the #include chain.
2. Should we generate bindings for system includes too? How we could control that?
3. Should we maintain the relative hierarchy observed whilst traversing #includes?

For item (3), I would say yes and I propose we should generate something like this:

app.scala
```scala
package com.example.app
...
```

codes.scala
```scala
package com.example.app
...
```

includes/shapes.scala
```scala
package com.example.app.includes
...
```

In other words, in this example in particular we do not generate bindings for system includes but we mimic the hierarchy of user includes:

```
/path/to/sources
  app.scala
  codes.scala
  + includes
      - shapes.scala
```

Maybe we could have a command line option which tells if we should mimic the hierarchy of includes or not. Since Scala does not care about files placed in specific directories which match the package name, the user may prefer to have all files generated in a single folder. The only relevant thing is how package names appear on these files. This could even work in case only one unified file is produced.
